conclave-enclave/src/main/kotlin/com/r3/conclave/enclave/Enclave.kt

```diff
@@ -150,7 +148,30 @@ abstract class Enclave {
     protected val signatureKey: PublicKey get() = signingKeyPair.public
 
     /** The serializable remote attestation object for this enclave instance. */
-    protected val enclaveInstanceInfo: EnclaveInstanceInfo get() = adminHandler.enclaveInstanceInfo
+    protected val enclaveInstanceInfo: EnclaveInstanceInfo by lazy {
+        val attestation = env.hostInterface.getAttestation()
+        val attestationReportBody = attestation.reportBody
+        val enclaveReportBody = getEnclaveInstanceInfoQuoteCallHandler.mostRecentQuote[quote][reportBody]
+
+        check(attestationReportBody == enclaveReportBody) {
+            """Host has provided attestation for a different enclave.
+               Expected: $enclaveReportBody
+               Received: $attestationReportBody""".trimIndent()
+        }
+
+        // It's also important to check the enclave modes match. Specifically we want to prevent an attestation marked
+        // as secure from being used when the enclave is running in non-hardware mode (all non-hardware attestations
+        // are insecure).
+        check(attestation.enclaveMode == env.enclaveMode) {
+            "The enclave mode of the attestation (${attestation.enclaveMode}) does not match ${env.enclaveMode}"
+        }
+
+        EnclaveInstanceInfoImpl(
+                signatureKey,
+                encryptionKeyPair.public as Curve25519PublicKey,
+                attestation
+        )
```

Comment:
There's some security checks here and thus the entire lazy block is better served to be its own separate method.